aim/agent/aid/universes/aci/aci_universe.py

```diff
@@ -113,37 +124,135 @@ def _reload_websocket_config(self):
         if set(ws_urls) != set(self.ws_urls):
             self.ws_urls = ws_urls
 
-    def establish_ws_session(self, max_retries=None):
+    def _update_apic_assign_db(self, aim_context, apic_assign,
+                               apic_assign_obj):
+        try:
+            if apic_assign_obj is None:
+                apic_assign.aim_aid_id = self.agent_id
+                return self.manager.create(aim_context, apic_assign)
+
+            if apic_assign_obj.aim_aid_id != self.agent_id:
+                obj = self.manager.update(aim_context, apic_assign_obj,
+                                          aim_aid_id=self.agent_id)
+            else:
+                # This will update the last_update_timestamp
+                # automatically
+                obj = self.manager.update(aim_context, apic_assign_obj)
+            return obj
+        # This means another controller is also adding/updating
+        # the same entry at the same time and he has beat us.
+        except Exception as e:
+            LOG.info(e)
+            return None
+
+    def _ws_session_login(self, url, url_max_retries, purpose,
+                          aim_context=None, apic_assign=None,
+                          apic_assign_obj=None):
+        retries = 0
+        LOG.info('Establishing %s WS connection with url: %s',
+                 purpose, url)
+        valid_session = None
+        while retries < url_max_retries:
+            session = acitoolkit.Session(
+                url, self.apic_username,
+                self.apic_password,
+                verify_ssl=self.verify_ssl_certificate,
+                cert_name=self.cert_name,
+                key=self.private_key_file)
+            resp = session.login()
+            if not resp.ok:
+                LOG.warn('%s Websocket connection failed: %s',
+                         purpose, resp.text)
+                retries += 1
+                if session.session:
+                    session.close()
+                continue
+            LOG.info('%s Websocket connection succeeded with url: %s',
+                     purpose, url)
+            valid_session = session
+            break
+
+        if valid_session:
+            # We don't need to claim the ownership if we are just
+            # picking up a backup APIC.
+            if purpose != BACKUP_PURPOSE:
+                obj = self._update_apic_assign_db(
+                    aim_context, apic_assign, apic_assign_obj)
```

Comment:
What if two different AIDs concurrently pick the same APIC URL, establish connections, and get to this point? Will the create() call in _update_apic_assign_db() fail for the 2nd one, or will the 2nd create() overwrite the 1st? Seems we need to make sure this AID actually ends up owning the APIC, and if it doesn't, pick a different URL and try again.

Reply:
In that case, the DB operation inside _update_apic_assign_db() will throw an exception which will be caught then None will be returned for the 2nd AID. Then it will go ahead to try the next URL.

Reply:
OK, I missed where the exception was being handled in _update_apic_assign_db(). I see that None will be returned and the loop will try the next URL.